torchmetrics/detection/map.py

```diff
@@ -279,20 +291,339 @@ def update(self, preds: List[Dict[str, Tensor]], target: List[Dict[str, Tensor]]
+        for k in range(K):
+            Nk = k * A * I
+            for a in range(A):
+                Na = a * I
+                for m, max_det in enumerate(self.max_detection_thresholds):
+                    # Load all image evals for current class_id and area_range
+                    E = [evalImgs[Nk + Na + i] for i in range(I)]
+                    E = [e for e in E if e is not None]
+                    if len(E) == 0:
+                        continue
+                    dt_scores = torch.cat([e["dtScores"][:max_det] for e in E])
+
+                    # different sorting method generates slightly different results.
+                    # mergesort is used to be consistent as Matlab implementation.
+                    inds = torch.argsort(dt_scores, descending=True)
+                    dt_scores_sorted = dt_scores[inds]
+
+                    dt_matches = torch.cat([e["dtMatches"][:, :max_det] for e in E], axis=1)[:, inds]
+                    dt_ignore = torch.cat([e["dtIgnore"][:, :max_det] for e in E], axis=1)[:, inds]
+                    gt_ignore = torch.cat([e["gtIgnore"] for e in E])
+                    npig = torch.count_nonzero(gt_ignore == False)  # noqa: E712
+                    if npig == 0:
+                        continue
+                    tps = torch.logical_and(dt_matches, torch.logical_not(dt_ignore))
+                    fps = torch.logical_and(torch.logical_not(dt_matches), torch.logical_not(dt_ignore))
+
+                    tp_sum = torch.cumsum(tps, axis=1, dtype=torch.float)
+                    fp_sum = torch.cumsum(fps, axis=1, dtype=torch.float)
+                    for t, (tp, fp) in enumerate(zip(tp_sum, fp_sum)):
+                        nd = len(tp)
+                        rc = tp / npig
+                        pr = tp / (fp + tp + torch.finfo(torch.float64).eps)
+                        q = torch.zeros((R,))
+                        ss = torch.zeros((R,))
+
+                        if nd:
+                            recall[t, k, a, m] = rc[-1]
+                        else:
+                            recall[t, k, a, m] = 0
+
+                        # Remove zigzags for AUC
+                        for i in range(nd - 1, 0, -1):
+                            if pr[i] > pr[i - 1]:
+                                pr[i - 1] = pr[i]
+
+                        inds = torch.searchsorted(rc, self.rec_thresholds, right=False)
+                        # TODO: optimize
+                        try:
+                            for ri, pi in enumerate(inds):  # range(min(len(inds), len(pr))):
+                                # pi = inds[ri]
+                                q[ri] = pr[pi]
+                                ss[ri] = dt_scores_sorted[pi]
+                        except Exception:
+                            pass
+                        precision[t, :, k, a, m] = q
+                        scores[t, :, k, a, m] = ss
```

Comment:
let's make this another private method...
the reason is that it is too complex and so harder to follow and eventually with using/calling internal functions we can increase correctness with intermediate typing check :rabbit: